### ISSUE TITLE
Feature Request: python-visible namespace for debugging

### ISSUE THREAD

**Issue by jbrockmendel** [cython/cython#1875]
I've spent untold hours in the following loop:

1) `python setup.py build_ext --inplace`
2) `import pandas as pd`
3) `ImportError: pandas/_libs/tslibs/_testing.so: undefined symbol: parse_iso_8601_datetime`
4) Yell profanity at my keyboard

So the feature I'm requesting is a way to determine what symbols _are_ defined.  Or near-equivalently: what is available for `cimport`?

For a specific example, take [this](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/hashtable.pyx#L40) extern from pandas._libs.hashtable:

```
cdef extern from "numpy/npy_math.h":
    double NAN "NPY_NAN"
```

I'd like to be able to accomplish this with `cimport`, and a way to figure out where to `cimport` it from.

I'll volunteer to put together a PR, need some guidance on where to start.

**Comment by robertwb** [comment]
What symbols are defined in where? Where do you expect
parse_iso_8601_datetime to be defined? Presumably if you're using it via
headers, those headers match the shared library you're linking against. And
if so you should be getting a compile time error or link error rather than
a runtime error.

Generally the set of cimportable symbols should be declared in a pxd
file(s) that ships with the library one defines (though for numpy in
particular we ship our own numpy.pxd so you can just do "from numpy cimport
NAN" or whatever).

FWIW, it's also recommended that one use cythonize so that when pxd files
change (either of an upstream library, or within your own code) all
dependent pyx files get recompiled rather than just the ones that were
touched.

(There's nm for an arbitrary library of course, but that's probably not
what you're looking for.)

**Comment by jbrockmendel** [comment]
> What symbols are defined in where? Where do you expect parse_iso_8601_datetime to be defined?

Trial and error with many variants and lots of very similar names means I'm going to have a hard time exactly replicating the exact errors from earlier.  Rather than give a kinda-sorta example, I'll hold onto this until next time this comes up and I can produce a real case.

<b>Update</b> Real life example: I cannot for the life of me figure out how to cimport `PyExc_ValueError`

> though for numpy in particular we ship our own numpy.pxd so you can just do "from numpy cimport NAN" or whatever

The "or whatever" bit is the part that could be made more python-friendly.  This might be an easy case of what I have in mind.  Imagine an interactive session:

```
>>> import cython
>>> dir(cython.numpy)
["import_array", "ndarray", "int64_t", ...]
```

> FWIW, it's also recommended that one use cythonize so that when pxd files [...]

Yah, there's been a couple of threads over at pandas recently regarding the specification of dependencies in the setup.py file.  The cython files seem pretty well-behaved in terms of cache-invalidation, but there are some .c and .h files that seem fragile.

(Incidentally, that's why I was tinkering with parse8601, trying to implement https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/datetime/np_datetime_strings.c#L377 in cython.  It came out something like 10x slower than the original, which probably indicates a good stretch of learning curve ahead)

**Comment by cython-notifications** [comment]
On Tue, Sep 19, 2017 at 9:47 PM, jbrockmendel <notifications@github.com>
wrote:

> What symbols are defined in where? Where do you expect
> parse_iso_8601_datetime to be defined?
>
> Trial and error with many variants and lots of very similar names means
> I'm going to have a hard time exactly replicating the exact errors from
> earlier. Rather than give a kinda-sorta example, I'll hold onto this until
> next time this comes up and I can produce a real case.
>
> *Update* Real life example: I cannot for the life of me figure out how to
> cimport PyExc_ValueError
>
You shouldn't ever need this from Cython, just use the builtin name
ValueError. (Was this an undefined symbol error? This should have been an
error at compile time if you tried to use it without it being defined).

> though for numpy in particular we ship our own numpy.pxd so you can just
> do "from numpy cimport NAN" or whatever
>
> The "or whatever" bit is the part that could be made more python-friendly.
> This might be an easy case of what I have in mind. Imagine an interactive
> session:
>
> >>> import cython
> >>> dir(cython.numpy)
> ["import_array", "ndarray", "int64_t", ...]
>
> Not quite following what you mean by cython.numpy.

Is this the set of symbols in numpy.pxd? (Generally pxd files are small
enough to just read, given they contain little to know implementation.) Or
the symbols defined in some header file? (Parsing C/C++ header files is,
well, an interesting exercise.)  Or from numpy.so (e.g. nm numpy.so)?

> FWIW, it's also recommended that one use cythonize so that when pxd files
> [...]
>
> Yah, there's been a couple of threads over at pandas recently regarding
> the specification of dependencies in the setup.py file. The cython files
> seem pretty well-behaved in terms of cache-invalidation, but there are some
> .c and .h files that seem fragile.
>
> (Incidentally, that's why I was tinkering with parse8601, trying to
> implement https://github.com/pandas-dev/pandas/blob/master/pandas/_
> libs/src/datetime/np_datetime_strings.c#L377 in cython. It came out
> something like 10x slower than the original, which probably indicates a
> good stretch of learning curve ahead)
>
Yeah, that'd probably be a tough one to replicate or beat starting from
scratch in plain old C.

**Comment by jbrockmendel** [comment]
> You shouldn't ever need [PyExc_ValueError] from Cython, just use the builtin name ValueError

The context of this was trying to match the original `parse_iso_8601_datetime` more closely.  It has:
```
parse_error:
    PyErr_Format(PyExc_ValueError,
                 "Error parsing datetime string \"%s\" at position %d", str,
                 (int)(substr - str));
    return -1;
```

So i tried defining:
```
cdef int parse_error(unsigned char* dstr, int position):
    raise ValueError("Error parsing datetime string \"%s\" at position %d",
                     <str>dstr,
                     position)
    #return -1
```
(Tried various other types for `dstr`, let's skip over that for the moment) This works fine until I try to tack `nogil` onto the definition, at which point I get compile-time errors `Converting to Python object not allowed without gil`.  So the thought was more closely matching the original might help.

> Yeah, that'd probably be a tough one to replicate or beat starting from
scratch in plain old C.

The actual goal (ease of packaging and dependency specification) would be accomplished if I could just drop the original C version into a pyx file.  Is that an option?

**Comment by scoder** [comment]
This discussion seems to derail a little, although it would make an excellent thread on the cython-users list. I would suggest to migrate it there, and if anything suitable comes out of it that motivates this ticket, copy the relevant comments back here.

**Comment by jbrockmendel** [comment]
Good call, will do.

**Comment by jbrockmendel** [comment]
> Not quite following what you mean by cython.numpy.

This part is actually pertinent to the original topic.  `cython.numpy` was an example, the idea being that it would basically just implement `__dir__` to make it easy for beginners to figure out from the interpreter what can be `cimport`ed.

With things like `np.get_include()` it can be non-obvious what is available in the namespace, and errors can pass silently.

An example of a silent namespace error: For quite a while I had a branch of [datetime.pxd](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/datetime.pxd) where I tried to extern-import "get_datetimestruct_days" as:

```
cdef extern from "numpy/ndarrayobject.h":
    npy_int64 get_datetimestruct_days(const npy_datetimestruct *dts)
```

... a name which I now know is not exported in that namespace (are "exported" and "namespace" the right terminology here?  Let me know if the meaning is not clear).  It did not cause an error because the same function _was_ exported by pandas' src/datetime/np_datetime.h, and while I still had a `cdef extern from "datetime/np_datetime.h":` block (even though `get_datetimestruct_days` was not referenced in this block).

To recap the original issue: it would be awesome to have a python-facing way of grokking namespaces.  I'll offer labor if there's an interest in making this happen.

**Comment by robertwb** [comment]
If you try to cimport something that doesn't exist, you'll get a Cython compile error. Likewise, if you try to cdef extern declare something that doesn't exist in the .c or .h files, you'll get a C compile error (unless of course the .h file declared a symbol but you didn't link against the library providing that symbol, but there's nothing you can do about that, that's how C works). 

I'm going to close this because it's still not clear what you're asking for. Feel free to take it up again on the list.
